networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/GossipMessageHandlerIntegrationTest.java

```diff
@@ -88,6 +105,116 @@ public void shouldGossipBlocksAcrossToIndirectlyConnectedPeers() throws Exceptio
+        new MockStartValidatorKeyPairFactory().generateKeyPairs(0, 12);
+    final BeaconChainUtil chainUtil = BeaconChainUtil.create(storageClient1, blsKeyPairList);
+    chainUtil.initializeStorage();
+
+    // Setup network 2
+    final EventBus eventBus2 = new EventBus();
+    final ChainStorageClient storageClient2 = new ChainStorageClient(eventBus2);
+    final JvmLibP2PNetwork network2 = networkFactory.startNetwork(eventBus2, storageClient2);
+    chainUtil.initializeStorage(storageClient2);
+
+    // Connect networks 1 -> 2
+    network1.connect(network2.getPeerAddress());
+    // Wait for connections to get set up
+    Waiter.waitFor(
+        () -> {
+          assertThat(network1.getPeerManager().getAvailablePeerCount()).isEqualTo(1);
+          assertThat(network2.getPeerManager().getAvailablePeerCount()).isEqualTo(1);
+        });
+    // TODO: debug this - we shouldn't have to wait here
+    Thread.sleep(2000);
+
+    // Listen for new block event to arrive on network 2
+    final AttestationCollector network2Attestations = new AttestationCollector(eventBus2);
+
+    // Propagate attestation from network 1
+    AttestationGenerator attestationGenerator = new AttestationGenerator(blsKeyPairList);
+    Attestation validAttestation = attestationGenerator.validAttestation(storageClient1);
+    eventBus1.post(validAttestation);
+
+    Thread.sleep(2000);
+
+    assertThat(network2Attestations.getAttestations()).isEmpty();
+  }
+
+  @Test
+  public void shouldReceiveAttestationAcrossPeersInSameSubnet() throws Exception {
+    final int numValidators = 12;
+
+    ArtemisConfiguration config = mock(ArtemisConfiguration.class);
+    doReturn(0).when(config).getNaughtinessPercentage();
+    doReturn(numValidators).when(config).getNumValidators();
+    doReturn(null).when(config).getValidatorsKeyFile();
+    doReturn(0).when(config).getInteropOwnedValidatorStartIndex();
+    doReturn(numValidators).when(config).getInteropOwnedValidatorCount();
+
+    AttestationAggregator aggregator = mock(AttestationAggregator.class);
+    doNothing().when(aggregator).updateAggregatorInformations(any());
+
+    // Setup network 1
+    final EventBus eventBus1 = new EventBus();
+    final ChainStorageClient storageClient1 = new ChainStorageClient(eventBus1);
+    final JvmLibP2PNetwork network1 = networkFactory.startNetwork(eventBus1, storageClient1);
+    List<BLSKeyPair> blsKeyPairList =
+        new MockStartValidatorKeyPairFactory().generateKeyPairs(0, numValidators);
+    final BeaconChainUtil chainUtil = BeaconChainUtil.create(storageClient1, blsKeyPairList);
+    chainUtil.initializeStorage();
+
+    // Setup network 2
+    final EventBus eventBus2 = new EventBus();
+    final ChainStorageClient storageClient2 = new ChainStorageClient(eventBus2);
+    final JvmLibP2PNetwork network2 = networkFactory.startNetwork(eventBus2, storageClient2);
+    chainUtil.initializeStorage(storageClient2);
+
+    // Setup VC 1
+    new ValidatorCoordinator(
```

Comment:
I think actually creating the `ValidatorCoordinator` may be increasing the scope of the integration test too far here.  In my mind I'm kind of using the event bus to define the boundaries for integration tests.  So in this integration test I'd just fire `eventBus.new CommitteeAssignmentEvent())` to get the network layer to join subnets and fire `CommitteeDismissalEvent` to leave them again and check it works by checking if `Attestation` events and seeing whether or not they broadcast.

We might have a separate integration test for `ValidatorCoordinator` to check that it generates the right events given specific inputs (though I think we can cover that with unit tests if the committee assignment/dismissal events are managed by a separate class as mentioned below).

Otherwise we're rapidly reaching the scope of an acceptance test and the test gets brittle because we're having to build up multiple components and hook them together just right without using the actual production code to do that.

And conceptually, components which only interact via the event bus should be very loosely coupled so should be primarily tested independently (and we can check it all hooks together with a few acceptance tests to check everything actually works together but that's outside the scope of this work).